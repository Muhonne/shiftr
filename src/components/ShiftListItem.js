// @flow

import React from "react";
import styled from "styled-components";
import {
  Animated,
  Easing,
  LayoutAnimation,
  TouchableWithoutFeedback
} from "react-native";

import Text from "./Text";
import Button from "./Button";
import type { Shift } from "../types";
import constants from "../constants";
import utils from "../utils";
import AnimatedView from "./AnimatedView";

const Container = styled.View`
  padding: ${constants.spacing.s}px;
  margin-top: ${constants.spacing.s}px;
  margin-left: ${constants.spacing.s}px;
  margin-right: ${constants.spacing.s}px;
  border: 1px;
  border-color: ${constants.colors.grey};
  border-radius: ${constants.spacing.s};
`;

const Row = styled.View`
  flex-direction: ${props => (props.open ? "column" : "row")};
  align-items: ${props => (props.open ? "flex-start" : "center")};
`;

const ContentRow = styled.View`
  margin-bottom: ${props => (props.open ? constants.spacing.s : 0)};
  flex-direction: row;
  align-items: center;
`;

const OpenContent = styled.View``;

const Icon = styled.Image`
  width: ${constants.fontSize.normal};
  height: ${constants.fontSize.normal};
  margin-right: ${constants.spacing.s};
`;

class ShiftListItem extends React.Component<
  { shift: Shift },
  { open: boolean, animate: any }
> {
  state = {
    open: false,
    animate: new Animated.Value(0)
  };

  toggle = () => {
    LayoutAnimation.easeInEaseOut();
    this.setState({ open: !this.state.open });
    if (this.state.open) {
      this.animate(0);
    } else {
      this.animate(1);
    }
  };

  animate = (toValue: number): void => {
    Animated.timing(this.state.animate, {
      toValue,
      duration: constants.animationDuration,
      easing: Easing.linear,
      useNativeDriver: true
    }).start();
  };

  interPolateValue = (start: number, end: number): any =>
    this.state.animate.interpolate({
      inputRange: [0, 1],
      outputRange: [start, end]
    });

  viewTransforms = [
    { scaleX: this.interPolateValue(1, 1.1) },
    { scaleY: this.interPolateValue(1, 1.1) },
    { translateX: this.interPolateValue(0, 5) }
  ];

  render() {
    const { shift } = this.props;
    const { open } = this.state;

    return (
      <TouchableWithoutFeedback onPress={this.toggle}>
        <Container>
          <Row open={this.state.open}>
            <AnimatedView width={50} transforms={this.viewTransforms}>
              <ContentRow open={this.state.open}>
                <Icon source={require("../img/place.png")} />
                <Text>{shift.area}</Text>
              </ContentRow>
            </AnimatedView>
            <AnimatedView width={50} transforms={this.viewTransforms}>
              <ContentRow open={this.state.open}>
                <Icon source={require("../img/time.png")} />
                <Text>{`${utils.formatDate(shift.startTime)}`}</Text>
              </ContentRow>
            </AnimatedView>
          </Row>
          {open && (
            <OpenContent>
              <Text>Shift ends at {utils.formatDate(shift.endTime)}</Text>
              <Text>
                Shift duration{" "}
                {utils.getDuration(shift.startTime, shift.endTime)}
              </Text>
              <Button
                onPress={() => {
                  console.log("keks");
                }}
              >
                <Text>Book/Cancel</Text>
              </Button>
            </OpenContent>
          )}
        </Container>
      </TouchableWithoutFeedback>
    );
  }
}
export default ShiftListItem;
