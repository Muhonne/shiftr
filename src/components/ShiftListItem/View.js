// @flow

import React from "react";
import styled from "styled-components";
import { Animated, Easing, TouchableWithoutFeedback } from "react-native";

import Text from "../Text";
import type { Shift } from "../../constants";
import constants from "../../constants";
import utils from "../../utils";
import ShiftButton from "../ShiftButton";
import LayoutAnimation from "../../LayoutAnimation";

const Container = styled(Animated.View)`
  padding: ${constants.spacing.s}px;
  margin-top: ${constants.spacing.s / 2}px;
  margin-bottom: ${constants.spacing.s / 2}px;
  margin-left: ${constants.spacing.s}px;
  margin-right: ${constants.spacing.s}px;
  border: 1px;
  border-color: ${constants.colors.grey};
  background-color: ${constants.colors.white};
`;

const Row = styled.View`
  flex-direction: row;
  align-items: center;
`;

const ContentRow = styled(Row)`
  margin-bottom: ${props => (props.open ? constants.spacing.s : 0)};
`;

const OpenContent = styled.View`
  flex: 0.5;
  margin: ${constants.spacing.s}px;
`;

const Icon = styled.Image`
  width: ${constants.fontSize.normal};
  height: ${constants.fontSize.normal};
  margin-right: ${constants.spacing.s};
`;

type Props = {
  shift: Shift,
  data?: Shift,
  refresh: () => void
};

export default class ShiftListItem extends React.Component<
  Props,
  { open: boolean, animate: any }
> {
  state = {
    open: false,
    animate: new Animated.Value(0)
  };

  toggle = () => {
    this.setState({ open: !this.state.open });
    if (this.state.open) {
      this.animate(0);
    } else {
      this.animate(1);
    }
  };

  animate = (toValue: number): void => {
    Animated.timing(this.state.animate, {
      toValue,
      duration: constants.animationDuration,
      easing: Easing.linear,
      useNativeDriver: true
    }).start();
  };

  interPolateValue = (start: number, end: number): any =>
    this.state.animate.interpolate({
      inputRange: [0, 1],
      outputRange: [start, end]
    });

  viewTransforms = [
    { scaleX: this.interPolateValue(1, 1.05) },
    { scaleY: this.interPolateValue(1, 1.05) }
  ];

  render() {
    // initial data from parent, override with data from shifts own endpoint
    const { id, area, startTime, endTime, booked } =
      this.props.data || this.props.shift;
    const { refresh } = this.props;
    const { open } = this.state;

    LayoutAnimation();

    return (
      <TouchableWithoutFeedback onPress={this.toggle}>
        <Container style={{ transform: this.viewTransforms }}>
          <Row open={this.state.open}>
            <ContentRow style={{ width: "45%" }} open={this.state.open}>
              <Icon source={require("../../img/place.png")} />
              <Text>{area}</Text>
            </ContentRow>
            <ContentRow style={{ width: "55%" }} open={this.state.open}>
              <Icon source={require("../../img/time.png")} />
              <Text>
                {`${utils.formatDate(startTime)}`}
                {open && ` -  ${utils.formatDate(endTime)}`}
              </Text>
            </ContentRow>
          </Row>
          {open && (
            <Row>
              <OpenContent>
                <Text>Shift duration</Text>
                <Text large>{utils.getDuration(startTime, endTime)}</Text>
              </OpenContent>
              <OpenContent>
                <ShiftButton
                  shiftId={id}
                  refreshParent={refresh}
                  parentBooked={booked}
                />
              </OpenContent>
            </Row>
          )}
        </Container>
      </TouchableWithoutFeedback>
    );
  }
}
