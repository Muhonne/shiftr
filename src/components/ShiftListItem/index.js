import { reduxAutoloader } from "redux-autoloader";

import apiCalls from "../../apiCalls";
import View from "./View";

export default reduxAutoloader({
  startOnMount: false,
  name: props => props.shift.id,
  apiCall: props => apiCalls.getShift(props.shift.id)
})(View);
