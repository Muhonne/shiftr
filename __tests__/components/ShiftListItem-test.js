import React from "react";
import renderer from "react-test-renderer";
import ShiftListItem from "../../src/components/ShiftListItem/View";

const shiftDate = new Date(2000, 1, 1, 1, 1);

describe("Button", () => {
  it("renders with minimun props", () => {
    expect(
      renderer.create(
        <ShiftListItem
          shift={{
            id: "test",
            startTime: shiftDate,
            endTime: shiftDate,
            booked: false
          }}
        />
      )
    ).toMatchSnapshot();
  });
});
