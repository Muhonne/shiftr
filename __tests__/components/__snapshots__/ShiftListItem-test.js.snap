// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders with minimun props 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#fff",
      "borderColor": "#cacaca",
      "borderStyle": "solid",
      "borderWidth": 1,
      "marginBottom": 5,
      "marginLeft": 10,
      "marginRight": 10,
      "marginTop": 5,
      "paddingBottom": 10,
      "paddingLeft": 10,
      "paddingRight": 10,
      "paddingTop": 10,
      "transform": Array [
        Object {
          "scaleX": 1,
        },
        Object {
          "scaleY": 1,
        },
      ],
    }
  }
  testID={undefined}
>
  <View
    open={false}
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        },
        undefined,
      ]
    }
  >
    <View
      open={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Array [
            Object {
              "marginBottom": 0,
            },
            Object {
              "width": "45%",
            },
          ],
        ]
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/img/place.png",
          }
        }
        style={
          Array [
            Object {
              "height": 16,
              "marginRight": 10,
              "width": 16,
            },
            undefined,
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "alignSelf": "flex-start",
              "color": "#000",
              "fontSize": 16,
            },
            undefined,
          ]
        }
      />
    </View>
    <View
      open={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          },
          Array [
            Object {
              "marginBottom": 0,
            },
            Object {
              "width": "55%",
            },
          ],
        ]
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/img/time.png",
          }
        }
        style={
          Array [
            Object {
              "height": 16,
              "marginRight": 10,
              "width": 16,
            },
            undefined,
          ]
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "alignSelf": "flex-start",
              "color": "#000",
              "fontSize": 16,
            },
            undefined,
          ]
        }
      >
        01.02.2000 1:01
      </Text>
    </View>
  </View>
</View>
`;
